src/cart/swim/swim.c

```diff
@@ -1060,18 +1046,17 @@ swim_progress(struct swim_context *ctx, int64_t timeout)
 		if (send_updates) {
 			rc = swim_updates_send(ctx, target_id, sendto_id);
 			if (rc) {
-				SWIM_ERROR("swim_updates_send(): "
-					   DF_RC"\n", DP_RC(rc));
+				SWIM_ERROR("swim_updates_send(): "DF_RC"\n", DP_RC(rc));
 				D_GOTO(out, rc);
 			}
 			send_updates = false;
 		} else if (now + 100 < ctx->sc_next_event) {
-			break;
+			break; /* break loop if need to wait more than 100 ms. */
 		}
```

Comment:
Since swim_progress is called (indirectly) from crt_progress, looping for 100 ms blocks any incoming requests and incoming replies, who may change the SWIM state. Would it be better to just return to crt_progress instead of looping inside swim_progress, given that swim has its own pthread and crt context now?

Reply:
This is exactly what is doing here. It will break the loop and return to crt_progress instead of waiting for next event in the loop.

Reply:
Here we only break the looping _after 100 ms_, no?

Reply:
Sorry, here we only break the looping _after reaching 100 ms before sc_next_event_, right?

Reply:
No. It break the loop if next event will be not soon. It checks estimation time at the end of each iteration and break instead of waist time inside next iteration. But if next event will be in 100 ms it continue loop and limits of provided timeout only.

Reply:
This is just **additional** exit from loop if nothing to do in nearest future. But it will exit anyway if timeout complete.

Reply:
The point is "soon" or "nearest future" can be about 100 ms. If swim_progress has no action to take, returning to crt_progress might be better than spinning for min(100 ms, timeout).

Reply:
If you conclude this is not a critical problem, I'd be fine for it to be addressed later.

Reply:
I cannot predict how long time it will out till the crt_progress call it again. So, here is trade of between exit immediately and potentially miss next event or wait in the loop and progress next event in time but eat more CPU in cycle.
Originally we always waits inside the loop until timeout complete and rely on external scheduler. This is intermediate solution. I think it's better to have it.

Reply:
OK. Fine to leave this to future work: We may still exceed the timeout passed to crt_progress.

Reply:
I think spinning 1/10 second is too much, fortunately the caller passes 1us as timeout so it wouldn't actually do it.

Reply:
ok. Let's refine this in next patch.